tests/non-js-resources/subtests/service-worker/webpack/lib/sw-info-plugin.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const {
+  sources: { ConcatSource },
+  util: { createHash },
+  Compilation,
+} = require('webpack');
+
+/** @typedef {import('webpack/lib/Compiler')} Compiler */
+
+module.exports = class ServiceWorkerInfoPlugin {
+  constructor({
+    test = file => file === 'sw.js',
+    filterAssets = () => true,
+  } = {}) {
+    this.test = test;
+    this.filterAssets = filterAssets;
+  }
+
+  /** @param {Compiler} compiler */
+  apply(compiler) {
+    compiler.hooks.thisCompilation.tap(
+      'ServiceWorkerInfoPlugin',
+      compilation => {
+        compilation.hooks.processAssets.tap(
+          {
+            name: 'ServiceWorkerInfoPlugin',
+            stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE,
+          },
+          assetsList => {
+            const assets = compilation.getAssets();
+            const swAssets = assets.filter(asset => this.test(asset.name));
+            const referencedAssets = assets
+              .filter(
+                asset =>
+                  this.filterAssets(asset.name) && !this.test(asset.name),
+              )
+              .sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
+
+            const { outputOptions } = compilation;
+            const versionHash = createHash(outputOptions.hashFunction);
+            if (outputOptions.hashSalt) {
+              versionHash.update(outputOptions.hashSalt);
+            }
+            for (const asset of referencedAssets) {
+              versionHash.update(asset.source.buffer());
+            }
+            const version = versionHash
+              .digest(outputOptions.hashDigest)
+              .slice(0, outputOptions.hashDigestLength);
+
+            const publicPath =
+              outputOptions.publicPath === 'auto'
```

Comment:
No that's not the same. `publicPath: ""` means no prefix, which basically resolves all assets relative to the page. It's equal to `publicPath: "./"`.

`publicPath: "auto"` determines the path from the url in the `<script>` tag. So it depends on runtime conditions.

Technically to handle `publicPath: "auto"` correctly in the service worker it would have to figure out it's own location e. g. from `self.location` can calculate the publicPath from that. But that complexity is not needed for this test case. And 99.9% users that use service worker know the publicPath at build time anyway.

It would make sense for use cases where you deploy the assets to an unknown location on the server. So sometimes at `/abc/` and sometimes at `/def/ghi/`.

We could also throw an exception that it's not implemented if this helps.

Reply:
Is this documented? I couldn't find it at https://webpack.js.org/guides/public-path/

Reply:
https://github.com/webpack/webpack.js.org/issues/4491